### DISCUSSION TITLE
Twitter + RSS + Markdown

### DISCUSSION THREAD

**Discussion by scripting** [scripting/reallysimple#6] [General]
First check out the [new features](http://scripting.com/2022/07/18/163658.html?title=twitterRssMarkdown) to support Markdown in RSS, along with hooking Twitter up to RSS. 

If you have any questions or comments, please post them here. 

Remember to be professional and respectful. :smile:

**Comment by jeffevans108** [comment]
not a question — since the get go, RSS is the Bling! tnx a gazillion Dave

**Reply by scripting** [reply]
@jeffevans108 -- keep on truckin! ;-)

**Comment by rsdoiel** [comment]
This is really inspiring, thank you. 

I need some advise on approach. I assemble by blog from Markdown documents. Some have frontmatter (i.e. a YAML block at the top of the document with metadata about the article). In your source:markdown example would you include the whole document or omit frontmatter?

**Comment by scripting** [comment]
@rsdoiel -- i'm glad you find this inspiring. I do too, esp the opportunity to work with smart people on something that matters. ;-)

I can't answer this question without more information.

Does your blog have an RSS feed? If so it would be helpful to see it and the source text with the YAML block.

**Reply by rsdoiel** [reply]
Here's the gory details that got me thinking about front matter and Markdown. You can see an example blog post
at 

```
---
title: "Installing Golang from source on RPi-OS for arm64"
series: "Raspberry Pi"
number: 1
author: "R. S. Doiel"
date: "2022-02-18"
copyright: "copyright (c) 2022, R. S. Doiel"
keywords: [ "raspberry pi", "Raspberry Pi OS", "arm64" ]
license: "https://creativecommons.org/licenses/by-sa/4.0/"
---


Installing Golang from Source on RPi-OS for arm64
==========================================

By R. S. Doiel, 2022-02-18

This are my quick notes on installing Golang from source on the Raspberry Pi OS 64 bit.

1. Get a working compiler
	a. go to https://go.dev/dl/ and download go1.17.7.linux-arm64.tar.gz
	b. untar the tarball in your home directory (it'll unpack to $HOME/go)
	c. `cd go/src` and `make.bash`
2. Move go directory to go1.17
3. Clone go from GitHub
4. Compile with the downloaded compiler
	a. `cd go/src`
	b. `env GOROOT_BOOTSTRAP=$HOME/go1.17 ./make.bash`
	c. Make sure `$HOME/go/bin` is in the path
	d. `go version`
```

The frontmatter is delimited by "---" (opening) and "---" (closing) alone on the line. This is common practice for people using Pandoc (me) and RMarkdown (colleagues at a small science institute where I work).

The section of the RSS I think would be updated is below (I truncated the copy I made from the RSS link above). Warning this looks really ugly, something I wasn't happy with but didn't have a better idea until I read your post on the source name spaces and what you've been up to with reallysimple.

```
<?xml version="1.0"?>
<rss version="2.0">
    <channel>
        <title>R. S. Doiel</title>
        <link>https://rsdoiel.github.io</link>
        <description>Robert&amp;#39;s ramblings and wonderigs</description>
        <pubDate>Tue, 19 Jul 2022 00:00:00 UTC</pubDate>
        <lastBuildDate>Tue, 19 Jul 2022 00:00:00 UTC</lastBuildDate>
        <generator>mkrss 1.0.2</generator>
        <item>
            <title>Installing Golang from source on RPi-OS for arm64</title>
            <link>/blog/2022/02/18/Installing-Go-from-Source-RPiOS-arm64.md</link>
            <description>&#xA;&#xA;Installing Golang from Source on RPi-OS for arm64&#xA;==========================================&#xA;&#xA;By R. S. Doiel, 2022-02-18&#xA;&#xA;This are my quick notes on installing Golang from source on the Raspberry Pi OS 64 bit.&#xA;&#xA;1. Get a working compiler&#xA;&#x9;a. go to https://go.dev/dl/ and download go1.17.7.linux-arm64.tar.gz&#xA;&#x9;b. untar the tarball in your home directory (it&#39;ll unpack to $HOME/go)&#xA;&#x9;c. `cd go/src` and `make.bash`&#xA;2. Move go directory to go1.17&#xA;3. Clone go from GitHub&#xA;4. Compile with the downloaded compiler&#xA;&#x9;a. `cd go/src`&#xA;&#x9;b. `env GOROOT_BOOTSTRAP=$HOME/go1.17 ./make.bash`&#xA;&#x9;c. Make sure `$HOME/go/bin` is in the path&#xA;&#x9;d. `go version` ...</description>
            <pubDate>Fri, 18 Feb 2022 00:00:00 UTC</pubDate>
            <guid>/blog/2022/02/18/Installing-Go-from-Source-RPiOS-arm64.md</guid>
        </item>
       <!-- ... and the rest of the feed follows -->
```

The YAML front matter seems like it should be snipped before if the primary source:markdown target is display but if the source:markdown is intended as an ingest field then it should include the whole document (because there is useful metadata about the post). That is where I started to scratch my head and realize I wasn't quite getting it.

Source urls --

Markdown version:
https://raw.githubusercontent.com/rsdoiel/rsdoiel.github.io/main/blog/2022/02/18/Installing-Go-from-Source-RPiOS-arm64.md

HTML verison:
https://rsdoiel.github.io/blog/2022/02/18/Installing-Go-from-Source-RPiOS-arm64.html

RSS: 
https://rsdoiel.github.io/rss.xml

All the best,

Robert

**Reply by scripting** [reply]
@rsdoiel -- thanks for providing the detail. 

The frontmatter is playing the role of RSS, so it should not be part of the RSS item. Just the text of the post itself, starting with "Installing Golang from Source..."

Some of the stuff in the frontmatter belongs at the top of your RSS feed, like the copyright.

You are repeating the title of the post in the RSS item and in the Markdown text. Most if not all RSS readers will repeat the title if you do it that way. Not a very good user experience. 

I guess the summary is that the YAML stuff is doing what RSS is doing...

THe best way to proceed imho is to try feeding your RSS into a typical feed reader and see what you get, and if you can, go back and revise the code that generates the RSS iteratively until you get something reasonable showing up in the reader.

**Reply by rsdoiel** [reply]
Thank you so much for the pointers! I'm rewriting the tools I use and this is a big help and inspiration.

**Comment by scripting** [comment]
BTW -- I just posted devnotes about Markdown in feeds.

http://scripting.com/2022/07/19/152235.html?title=devNotesForMarkdownInFeeds

**Reply by colin-walker** [reply]
Love all this @scripting – I finally got round to adding markdown support to my feed. Also, my own reader integrated with my blog checks for source:markdown and uses that if present (your FeedLand feed has been the perfect test feed 😄). The more flexibility we have with RSS the better for the web as a whole. Keep up the good work 👍

**Reply by scripting** [reply]
@colin-walker -- excellent. Thanks for the support for markdown and thanks for the encouragement. It seems like things are moving again.

**Reply by scripting** [reply]
@colin-walker -- if your reader is somewhere public, i'd love to have a look.

**Reply by colin-walker** [reply]
@scripting It's public/private: it's a full reader for me when logged in and analogous to a blog roll (with reader) for everyone else showing posts from each feed within the last month.

You can see it [here](https://colinwalker.blog/reader/).

[Here's a post](https://colinwalker.blog/blog/?date=2022-04-08#p2) I wrote about it back in April - it's evolved a bit since.

I wanted a reader that integrated directly with my blog so I could perform indieweb functions such as like, reply & bookmark directly in the reader and have that then posted to the blog automatically.

**Reply by scripting** [reply]
@colin-walker -- that is really pretty, and i love the way my markdown text looks in there. 

i'm looking around your neighborhood -- have to work together. 

have you looked at feedland.org yet? 

given what i've seen i think you really will. 

there's a lot of new functionality there, i've been working on only this for almost a year and some of the ideas go back over a decade that i just got around to implementing like checkbox news. 

given what you're doing i think you would get it all right off the bat.

**Reply by colin-walker** [reply]
I don't have a Twitter account so can't log in but I've been reading all your posts about it and it looks really good. I definitely get what you're doing.

Being able to post within it and generate a feed (rather than just consume) is great and goes back to what Tantek Çelik said years ago about why social networks became so prevalent: you could read and write simply within the same interface.

Micro.blog has been about the only other place that has done anything like that so FeedLand is sorely needed. Especially with the current Twitterpocalypse.

**Reply by scripting** [reply]
@colin-walker -- create a twitter account, don't use your real name. you really should see it. i want your ideas on how to hook up to the stuff you're doing, but if you don't know you can't do it. 

you could start by reading the docs if you want. 

http://docs.feedland.org/firstThings.opml

all the other pages are linked in the left margin.

if you want I'll give you one of my many test twitter accounts.

**Reply by colin-walker** [reply]
@scripting I'm in and digging around a bit, will have more time later after work. Turning feeds (generally singular) into a community activity is great. There is the obvious worry over popularity contests should new users only pick those feeds with the most subscribers but having alternative views like the subscription log can help to offset this to a degree.

Discovery is always one of the hardest things to bootstrap but a community can help as everyone's experience of the web is different. It really is blogrolls on steroids.

On a side note, I was looking at rssCloud. It's probably not needed for most feeds (real time updates is unlikely to matter in the majority of situations) but it's definitely a nice to have for some. If it helps offset polling needs, however, that can only be a good thing.

I was doing a test against rpc.rsscloud.io and was obviously doing something wrong as I wasn't getting any response to ping attempts from PHP. That's going to be my lack of knowledge rather than any issues as it's obviously working fine 😅

**Reply by scripting** [reply]
@colin-walker -- I started a thread on rssCloud. I'm talking with other people about this, we might as well help you and other people at the same time.

https://github.com/scripting/reallysimple/issues/11

**Reply by colin-walker** [reply]
I *think* I've got it. My ignorance getting the better of me.

**Reply by scripting** [reply]
OK if you have questions, post them in the thread i pointed to above.

**Reply by scripting** [reply]
And if you get it working, please post an url so i can subscribe.

**Reply by colin-walker** [reply]
Seemed to work okay, the log showed the ping and it recognised an update 🤞

The feed is [here](https://colinwalker.blog/livefeed.xml).

I'm going to be blogging properly again from the 25th so you'll have to let me know if it is all okay.

**Reply by scripting** [reply]
If you post this in the thread I started more people who can help will see it. I’m not the expert on this.

**Comment by josephzitt** [comment]
One question: It's very cool that you're supporting Markdown -- but which Markdown are you using? There appear to be a lot of variants, many of which think that they are the One True Markdown. When you discuss it in the blog, could you link to a specification or something, to avoid confusion (such as mine)?

**Comment by scripting** [comment]
@josephzitt -- of course as you know there is no One True Markdown. 

I think the approach we all take with Markdown text is to stay within a reasonable subset of the syntax, and hope for the best. ;-)

I go by this [cheat sheet](https://www.markdownguide.org/cheat-sheet/), which I keep in an icon in the chrome of my browser, easy to reach for. And it seems to work. 

Re the Markdown processors I use --

1. On the client side -- I use something called [MarkdownCoverter](http://fargo.io/code/markdownConverter.js), which I downloaded many years ago. I seem to recall it's the one that StackExchange uses, so I figured that's got to be pretty good.

2. On the server side -- I use [markd](https://www.npmjs.com/package/marked). I don't recall how I decided to use this one, but it has served me well.

As I understand it Gruber like the way it has evolved. The shit does kind of hit the fan when it comes to using Markdown in something like RSS where interop is the primary driving force. But this was one of the first enhancements I wanted to get out there in 2022, the 20th year of RSS 2.0. :smile:

PS: I made this into a [post](http://scripting.com/2022/07/20/135214.html?title=whichMarkdown) on my blog.

**Comment by pzingg** [comment]
@scripting FYI I just submitted a pull request in the Akkoma (Mastodon-compatible server) repository to add `<source:markdown>` to their RSS feeds:  https://akkoma.dev/AkkomaGang/akkoma/pulls/343

If I have time, I may try to do the same for glitch-soc. I saw you have a GitHub issue there.

I also have a question on how Mastodon, as a feed producer, should set the `<cloud>` element. I see that the way WordPress.com implements it, the rssCloud server is set to the same domain as the originating WordPress blog (they must have implemented their own rssCloud server). In the Mastodon server case, do you think that the rssCloud server domain should be a single setting for the entire server (some of these servers have 500K users), or a per-user (per-feed) preference? 

I can probably hack Akkoma into supporting rssCloud, by implementing the functionality of Andrew Shell's example in Elixir. Then there could be a server-wide setting to use either Akkoma's own implementation, or any other public server (like rpc.rsscloud.org if it's still active).

**Comment by scripting** [comment]
@pzingg --

thanks for putting out the word on souce:markdown support. once we get a little critical mass there it should get really interesting. ;-)

i would much prefer it not be a feed-level preference setting, and that mastodon servers include their own cloud server support. 

or we could see what @andrewshell thinks -- i don't run a cloud server myself, i just use his.

**Reply by andrewshell** [reply]
Yeah, I’m guessing if a substantial number of servers start pinging my instance it will probably break. Best case is if is handled by mastodon. It’s a super simple protocol and what mastodon is doing already is probably pretty similar.

**Reply by pzingg** [reply]
Sadly I doubt the Akkoma gang will accept my PR. 

In submitting the PR I have some questions for you about recommendations for escaping characters like quotes, ampersands, angle brackets, newlines and “&amp;”-like entities in RSS. Would the escaping be the same for HTML and Markdown? I found a bunch of articles from 2008 to the present that seem to indicate that feed readers don’t  unescape content (title, description) the same way. Are there best practices?

Also I am making progress in hooking up RSSCloud “ping” messages to Akkoma status updates in real time. Did you know that Mastodon API has a websocket stream for updates?

https://docs.joinmastodon.org/methods/streaming/#web-socket

I was able to listen to the public firehose of updates for an Akkoma server with the “public:local” stream. Now to filter those updates by subscribed user feeds and then send a ping to the little Elixir RSSCloud server I’m building. All this can be done outside of the Mastodon/Akkoma server.

**Comment by scripting** [comment]
@pzingg -- I recommend using my daverss package as prior art. 

Do what it does. I never get complaints about it not working with feed readers.

https://github.com/scripting/rss

**Comment by pzingg** [comment]
I just today added three @scripting inspired issues to Akkoma (a Mastodon API-compatible server), and I can try to do the same for either Mastodon or one of its forks that supports Markdown, namely [glitch-soc](https://github.com/glitch-soc/mastodon). The main Mastodon repository still only support plain text in posts, whereas glitch-soc also supports posts authored in HTML and Markdown.

If you have any more firepower to add to the "Idea" or "Reasoning" sections of these issues, please expand here, and I'll amend the issues. I have pull requests ready to go for all these items...

* [Support RSS Cloud](https://akkoma.dev/AkkomaGang/akkoma/issues/389)
* [Support title-less feed items](https://akkoma.dev/AkkomaGang/akkoma/issues/391)
* [Support round-trip Markdown (and other content types)](https://akkoma.dev/AkkomaGang/akkoma/issues/392)
